horovod/runner/mpi_run.py

```diff
@@ -151,51 +164,71 @@ def mpi_run(settings, nics, env, command, stdout=None, stderr=None):
     mpi_ssh_args = ''
     if ssh_args:
         joined_ssh_args = ' '.join(ssh_args)
-        mpi_ssh_args = f'-mca plm_rsh_args \"{joined_ssh_args}\"'
-
-    # if user does not specify any hosts, mpirun by default uses local host.
-    # There is no need to specify localhost.
-    hosts_arg = '-H {hosts}'.format(hosts=settings.hosts)
+        mpi_ssh_args = f'-bootstrap=ssh -bootstrap-exec-args \"{joined_ssh_args}\"' if impi else f'-mca plm_rsh_args \"{joined_ssh_args}\"'
```

Comment:
All these `if impi else` should go behind some kind of MPI implementation args interface that provides these dialects. This becomes hard to read already, with the next MPI implementation that has different parameters for these things, it will become unreadable.

That api would also include the `mpi_impl_flags` and `impl_binding_args`. So `_get_mpi_implementation_flags` should read `_get_mpi_command`, which provides the command string and internalizes all flags and bindings and dialect specific strings.

@tgaddair what are your thoughts?

Reply:
I agree, a refactor so that every MPI implementation is its own class would greatly simplify things.